CHG: switch to ruff format instead of black entirely

sbrugman opened the pull request:
## Description

Recommend `ruff format` instead of `black`
Benefits performance as well as fewer dependencies that can conflict (e.g. `click`)

Link to ruff's pre-commit version to avoid having to update the docs every time there is a new release, which are frequent.

<!-- Why was this PR created? -->

## Development notes
<!-- What have you changed, and how has this been tested? -->


## Developer Certificate of Origin
We need all contributions to comply with the [Developer Certificate of Origin (DCO)](https://developercertificate.org/). All commits must be signed off by including a `Signed-off-by` line in the commit message. [See our wiki for guidance](https://github.com/kedro-org/kedro/wiki/Guidelines-for-contributing-developers/).

If your PR is blocked due to unsigned commits, then you must follow the instructions under "Rebase the branch" on the GitHub Checks page for your PR. This will retroactively add the sign-off to all unsigned commits and allow the DCO check to pass.

## Checklist

- [X] Read the [contributing](https://github.com/kedro-org/kedro/blob/main/CONTRIBUTING.md) guidelines
- [X] Signed off each commit with a [Developer Certificate of Origin (DCO)](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/managing-repository-settings/managing-the-commit-signoff-policy-for-your-repository)
- [X] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [X] Updated the documentation to reflect the code changes
- [X] Added a description of this change in the [`RELEASE.md`](https://github.com/kedro-org/kedro/blob/main/RELEASE.md) file
- [X] Added tests to cover my changes
- [X] Checked if this change will affect Kedro-Viz, and if so, communicated that with the Viz team

astrojuanlu commented on docs/source/development/linting.md:
```diff
@@ -9,34 +9,37 @@ Linting tools check your code for errors such as a missing bracket or line inden
+There are a variety of Python tools available to use with your Kedro projects. This guide shows you how to use [`ruff`](https://beta.ruff.rs).
+
+**`ruff`** is a fast linter and formatter that replaces `flake8`, `pylint`, `pyupgrade`, `isort`, `black` and [more](https://beta.ruff.rs/docs/rules/).
+  - It helps to make your code compliant to [`pep8`](https://pypi.org/project/pep8/).
```
```suggestion
  - It helps to make your code compliant to [PEP 8](https://peps.python.org/pep-0008/).
```

astrojuanlu commented on docs/source/development/linting.md:
```diff
@@ -9,34 +9,37 @@ Linting tools check your code for errors such as a missing bracket or line inden
+**`ruff`** is a fast linter and formatter that replaces `flake8`, `pylint`, `pyupgrade`, `isort`, `black` and [more](https://beta.ruff.rs/docs/rules/).
+  - It helps to make your code compliant to [`pep8`](https://pypi.org/project/pep8/).
+  - It reformats code and sorts imports alphabetically and automatically separating them into sections by
+type. You can read more in the [`black` documentation](https://black.readthedocs.io/en/stable/).
```
It's weird to link to the Black and isort docs when talking about ruff, does ruff have a page explaining this by any chance? Otherwise I'd even remove these links

stichbury replied:
Seconded. Is this just left over from the original docs in error?

astrojuanlu commented on docs/source/development/linting.md:
```diff
@@ -72,36 +76,44 @@ These hooks are run before committing your code to your repositories to automati
   - repo: https://github.com/astral-sh/ruff-pre-commit
     # Ruff version.
-    rev: v0.0.270
+    rev: '' # Replace with latest version
```
```suggestion
    rev: '' # Replace with latest version, for example 'v0.1.8'
```

astrojuanlu reviewed (approved):
🚀 Yes! Approving this with some nitpicks

stichbury commented on docs/source/development/linting.md:
````diff
@@ -72,36 +76,44 @@ These hooks are run before committing your code to your repositories to automati
+      - id: ruff-format
 ```
+
+The latest configuration for `ruff`'s pre-commit is available [here](https://github.com/astral-sh/ruff-pre-commit).
````
See Vale message. We don't like to put links under the word "here" as it's vague and not very accessible for people using screen readers. Can I suggest you reword to "See GitHub for [the latest configuration for ruff's pre-commit]." and link as I've done?

astrojuanlu commented:
https://github.com/kedro-org/kedro/issues/3395 is getting annoying...

merelcht commented:
I really like this suggestion, but we should then also update the behaviour of our `lint` tool, because that still uses `black`. As a user I would be confused that `ruff format` is the recommendation in the docs, but the tool we provide has a different setup.

sbrugman commented:
@merelcht Agreed, better to switch to `ruff` for both linting and formatting completely. 
`bandit` is now also replaced with `ruff`

I found some obviated `noqa` comments and removed them.

It would be good to create a follow-up story that expands the selection of `ruff` rules, since we only use a subset of the rich library

merelcht commented on RELEASE.md:
```diff
@@ -6,6 +6,7 @@
 * Removed example pipeline requirements when examples are not selected in `tools`.
 * Allowed modern versions of JupyterLab and Jupyter Notebooks.
 * Removed setuptools dependency
+* Removed black dependency
```
```suggestion
* Removed black dependency and replaced it functionality with `ruff format`.
```

merelcht commented on tests/framework/cli/test_starters.py:
```diff
@@ -132,17 +132,19 @@ def _assert_requirements_ok(
-        assert "black" in requirements
         assert "ruff" in requirements
 
+        print(pyproject_file_path.read_text())
         pyproject_config = toml.load(pyproject_file_path)
+        print(pyproject_config)
```
I'm guessing these `print` statements are left overs from debugging?

merelcht commented:
The tests are currently failing because the `pyproject.toml` files in [kedro-starters](https://github.com/kedro-org/kedro-starters/tree/main) need to be updated to contain all `ruff` setup as well.

merelcht commented:
> The tests are currently failing because the `pyproject.toml` files in [kedro-starters](https://github.com/kedro-org/kedro-starters/tree/main) need to be updated to contain all `ruff` setup as well.

Created a PR for this https://github.com/kedro-org/kedro-starters/pull/207

noklam reviewed (approved):
🙏Appreciate your help!

SajidAlamQB commented:
Added a change so that unit tests now use the `main` branch of kedro-starters its only failing on coverage for `thread_runner.py`.

merelcht commented on kedro/framework/session/session.py:
```diff
@@ -33,20 +33,20 @@ def _describe_git(project_path: Path) -> dict[str, dict[str, Any]]:
-            ["git", "rev-parse", "--short", "HEAD"],
-            cwd=path,
+            ["git", "rev-parse", "--short", "HEAD"],  # noqa: S603, S607
+            cwd=project_path,
```
```suggestion
            cwd=path,
```

merelcht commented on kedro/framework/session/session.py:
```diff
@@ -33,20 +33,20 @@ def _describe_git(project_path: Path) -> dict[str, dict[str, Any]]:
-            ["git", "status", "--short"],
-            cwd=path,
+            ["git", "status", "--short"],  # noqa: S603, S607
+            cwd=project_path,
```
```suggestion
            cwd=path,
```
